File: src/main/java/tech/igrant/no46/Solution.java
package tech.igrant.no46;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

class Solution {

    public List<List<Integer>> permute(int[] nums) {
        return generate(Arrays.stream(nums).boxed().collect(Collectors.toList()));
    }

    List<List<Integer>> generate(List<Integer> array) {
        if (array.size() == 1) {
            return Collections.singletonList(array);
        }
        final int[] holder = new int[]{0};
        return array.stream().
                map(curr -> generate(remove(holder[0]++, array))
                        .stream()
                        .map(ret -> merge(ret, curr))
                        .collect(Collectors.toList())
                )
                .reduce(
                        new ArrayList<>(),
                        (acc, curr) -> {
                            acc.addAll(curr);
                            return acc;
                        }
                );
    }

    private List<Integer> merge(List<Integer> ret, Integer integer) {
        ret.add(integer);
        return ret;
    }

    List<Integer> remove(int index, List<Integer> origin) {
        List<Integer> ret = new ArrayList<>(origin);
        ret.remove(index);
        return ret;
    }

}

File: src/main/java/tech/igrant/no46/Solution.kt
package tech.igrant.no46

import java.util.stream.Collectors

class SolutionKt {

    fun permute(array: IntArray): List<List<Int>> {
        return if (array.size == 1) {
            listOf(listOf(array[0]))
        } else {
            array.indices
                    .map { index: Int ->
                        permute(remove(index, array))
                                .stream()
                                .map { ret -> ret + array[index] }
                                .collect(Collectors.toList())
                    }
                    .toList()
                    .flatten()
        }
    }

    private fun remove(index: Int, array: IntArray): IntArray {
        return IntArray(array.size - 1).apply {
            System.arraycopy(array, 0, this, 0, index)
            System.arraycopy(array, index + 1, this, index, this.size - index)
        }
    }

}